pkg/resourceoverrides/alloydb_instance.go

```diff
@@ -0,0 +1,161 @@
+// Copyright 2024 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package resourceoverrides
+
+import (
+	"fmt"
+	"strings"
+
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/k8s"
+	"github.com/GoogleCloudPlatform/k8s-config-connector/pkg/util/crdutil"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/terraform"
+
+	apiextensions "k8s.io/apiextensions-apiserver/pkg/apis/apiextensions/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+)
+
+var (
+	instanceTypeField    = []string{"instanceType"}
+	instanceTypeRefField = []string{"instanceTypeRef"}
+)
+
+func GetAlloyDBInstanceResourceOverrides() ResourceOverrides {
+	ro := ResourceOverrides{
+		Kind: "AlloyDBInstance",
+	}
+	// Preserve the legacy non-reference field 'instanceType' after it is changed to
+	// a reference field, 'instanceTypeRef'.
+	ro.Overrides = append(ro.Overrides, keepInstanceTypeField())
+	return ro
+}
+
+func keepInstanceTypeField() ResourceOverride {
+	o := ResourceOverride{}
+	o.CRDDecorate = func(crd *apiextensions.CustomResourceDefinition) error {
+		return preserveBothInstanceTypeAndInstanceTypeRefInCRD(crd, instanceTypeRefField[0], instanceTypeField[0])
+	}
+	o.PostActuationTransform = func(original, reconciled *k8s.Resource, tfState *terraform.InstanceState, dclState *unstructured.Unstructured) error {
+		if err := PreserveUserSpecifiedLegacyField(original, reconciled, instanceTypeField...); err != nil {
+			return fmt.Errorf("error preserving '%v' in post-actuation transformation: %w", strings.Join(instanceTypeField, "."), err)
+		}
+		if err := PruneDefaultedAuthoritativeFieldIfOnlyLegacyFieldSpecified(original, reconciled, instanceTypeField, instanceTypeRefField); err != nil {
+			return fmt.Errorf("error conditionally pruning '%v' in post-actuation transformation: %w", strings.Join(instanceTypeRefField, "."), err)
+		}
+		return nil
+	}
+	return o
+}
+
+func preserveBothInstanceTypeAndInstanceTypeRefInCRD(crd *apiextensions.CustomResourceDefinition, referenceFieldName, nonReferenceFieldName string) error {
```

Comment:
AI for KCC team: We should probably support a util function for multiple versions.